src/context/node.js

```diff
@@ -0,0 +1,414 @@
+      if (getMode().localDev || getMode().test) {
+        // The `Object.defineProperty({enumerable: false})` helps tests, but
+        // hurts performance. So this is only done in a dev/test modes.
+        Object.defineProperty(node, NODE_PROP, {
+          value: contextNode,
+          writable: false,
+          enumerable: false,
+          configurable: false,
+        });
+      } else {
+        node[NODE_PROP] = contextNode;
+      }
+    }
+    return contextNode;
+  }
+
+  /**
+   * Returns the closest available context node to the one specified. If the
+   * `node` has the context node, it's returned unless `excludeSelf` is set as
+   * `true`.
+   *
+   * The DOM traversal goes at most as far as the root node (document,
+   * shadow root, document fragment) or as far as the DOM tree allows. The
+   * traversal follows the assigned shadow slots.
+   *
+   * Root nodes (document or shadow root) and AMP elements are auto-created
+   * during the traversal.
+   *
+   * @param {!Node} node The node from which to perform the search.
+   * @param {boolean=} excludeSelf Whether the specified node itself should
+   * be excluded from the search. Defaults to `false` - do not exclude.
+   * @return {?ContextNode}
+   */
+  static closest(node, excludeSelf = false) {
+    let n = node;
+    while (n) {
+      // Check if a node is a candidate to be returned.
+      if (n != node || !excludeSelf) {
+        if (n[NODE_PROP]) {
+          // Already a discovered node.
+          return /** @type {!ContextNode} */ (n[NODE_PROP]);
+        }
+        const {nodeType} = n;
+        if (nodeType == DOCUMENT_NODE || nodeType == FRAGMENT_NODE) {
+          // A context node is always created for a root. Due to this, a
+          // non-root element is always at least attached to a root. This
+          // allows for quick discovery and reattachment when new information
+          // becomes available.
+          return ContextNode.get(n);
+        }
+        if (nodeType == ELEMENT_NODE && startsWith(n.tagName, AMP_PREFIX)) {
+          // An AMP node will always have a context node backing it at some
+          // point.
+          return ContextNode.get(n);
+        }
+      }
+      // Navigate up the DOM tree. Notice that we do not automatically go over
+      // a root node boundary.
+      const assignedSlot =
+        /** @type {?Node|undefined} */ (n[ASSIGNED_SLOT_PROP]) ||
+        n.assignedSlot;
+      if (assignedSlot) {
+        n = assignedSlot;
+      } else {
+        n = n.parentNode;
+      }
+    }
+    return null;
+  }
+
+  /**
+   * Direct slot assignment. Works the same way as shadow slots, but does not
+   * require a shadow root. Automatically starts the discovery phase for the
+   * affected nodes.
+   *
+   * See `Element.assignedSlot` API.
+   *
+   * @param {!Node} node The target node.
+   * @param {!Node} slot The slot to which the target node is assigned.
+   */
+  static assignSlot(node, slot) {
```

Comment:
Why do we need slot methods? Shouldn't normal ancestry lookup work fine?

Reply:
Normal ancestry doesn't work here. So this code reconstructs the composed tree. Couple of ways to think about this:

1. One model is [`composed` event propagation](https://developer.mozilla.org/en-US/docs/Web/API/Event/Event). This tree follows the path of a composed event. E.g. a click on a distributed node will first go to a slot, then shadow root, and only then to the light-dom parent.
2. Another model is style tree, which is also roughly a composed tree. E.g. by default, as soon as the shadow root is attached, all children are undisplayed. When an element is distributed into a slot, it becomes displayed as the slot and slot parent is displayed. E.g. a carousel's shadow tree container defines `display: flex` and the element's light-dom tree become "flex items".

Following this model, both event-style bubble propagation (child->parent) and broadcast (parent -> child) are better over the composed tree. A simple example: if a carousel component defines `playable=false` on a slide (a slot), this state should propagate to the distributed subtree of this slide and also make it `playable=false`. However, playable state of the light-dom parent (the carousel itself) will not changed here.